Sources/SPTPersistentCacheGarbageCollectorScheduler.m

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2016 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#import "SPTPersistentCacheGarbageCollectorScheduler.h"
+#import "SPTPersistentCacheTypeUtilities.h"
+#import "SPTPersistentCache+Private.h"
+
+BOOL SPTPersistentCacheGarbageCollectorSchedulerIsInMainQueue(void);
+
+static const NSTimeInterval SPTPersistentCacheTimerProxyTimerToleranceInterval = 300;
+
+@interface SPTPersistentCacheGarbageCollectorScheduler ()
+@property (nonatomic, strong) SPTPersistentCacheDebugCallback debugOutput;
+@property (nonatomic, strong) NSTimer *timer;
+@property (nonatomic, strong) SPTPersistentCacheOptions *options;
+@end
+
+
+@implementation SPTPersistentCacheGarbageCollectorScheduler
+
+#pragma mark - Initializer
+
+- (instancetype)initWithDataCache:(SPTPersistentCache *)dataCache
+                          options:(SPTPersistentCacheOptions *)options
+                            queue:(dispatch_queue_t)queue
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+    
+    _options = options;
+    _dataCache = dataCache;
+    _queue = queue;
+    _debugOutput = options.debugOutput;
+    
+    return self;
+}
+
+- (void)dealloc
+{
+    NSTimer *timer = _timer;
```

Comment:
Sorry, the retain-cycle problem is still here unfortunately in the generic case. As if the timer hasn’t been invalidated you’ll never reach this method. And if it has been invalidated already your implementation doesn’t do anything.

So do **one** of the following two solutions. Of which I’d prefer option number two as it’s the most resilient with the least likelihood of introducing retain-cycles due to human error.

## Solution one, let external users deal with it
Document that `-unscheduleGarbageCollection` must be sent to the scheduler or it will lead to horrors. Then remove the entire `-dealloc` implementation as it will never be called when a timer is alive, and as such it can’t invalidate said timer.

## Solution two, add a weak proxy to deal with it automatically
Add a `NSProxy` subclass called `SPTPersistentCacheWeakTargetProxy` and use it as the timer’s target:
```objc
- (void)scheduleGarbageCollection
{
    // ...

    self.timer = [NSTimer timerWithTimeInterval:self.options.gcIntervalSec
                                         target:[SPTPersistentCacheWeakTargetProxy proxyWithTarget:self]
                                       selector:@selector(enqueueGarbageCollection:)
                                       userInfo:nil
                                        repeats:YES];

    // ...
}
```

The implementation of `SPTPersistentCacheWeakTargetProxy` would look like:
```objc
// sources/SPTPersistentCacheWeakTargetProxy.h

NS_ASSUME_NONNULL_BEGIN

/**
 * A proxy class which insulates it’s target from being strongly retained. It will forward any messages sent
 * to itself to its target.
 *
 * Can be used together with `NSTimer` to make sure a retain-cycle isn’t created.
 */
@interface SPTPersistentCacheWeakTargetProxy : NSProxy

/**
 * Create a new weak target proxy for the given _target_.
 *
 * @param target The target to which all received messages should be forwarded.
 */
+ (instancetype)proxyWithTarget:(id)target;

@end

NS_ASSUME_NONNULL_END


// sources/SPTPersistentCacheWeakTargetProxy.m
@interface SPTPersistentCacheWeakTargetProxy ()

@property (nonatomic, weak, readonly) id target;

@end

@implementation SPTPersistentCacheWeakTargetProxy

#pragma mark Object Life Cycle

+ (instancetype)proxyWithTarget:(id)target
{
    return [[self alloc] initWithTarget:target];
}

- (instancetype)initWithTarget:(id)target
{
    _target = target;
    return self;
}

#pragma mark Forwarding Messages

- (id)forwardingTargetForSelector:(SEL)aSelector
{
    return self.target;
}

- (NSMethodSignature *)methodSignatureForSelector:(SEL)selector
{
    id target = self.target;

    NSMethodSignature *methodSignature = [object methodSignatureForSelector:selector];

    if (methodSignature == nil) {
        // We still want to report to the runtime that we found a selector. It’ll just be sent to `nil` later
        // so not a problem.
        NSString * const types = [NSString stringWithFormat:@"%@%@", @(@encode(id)), NSStringFromSelector(@encode(SEL))];
        methodSignature = [NSMethodSignature signatureWithObjCTypes:types.UTF8String];
    }
    return methodSignature;
}

- (void)forwardInvocation:(NSInvocation *)anInvocation
{
    if (self.target) {
        [anInvocation invokeWithTarget:self.target];
    }
}

@end
```